### PR TITLE
Fix problem resolving bindgen.properties in a path containing spaces

### PR DESCRIPTION
**Pull request by nfekete**
This commit fixes https://github.com/igloo-project/bindgen/issues/12

I replaced the URI + File based code with just URIs when trying to
resolve bindgen.properties. The code works by creating new URIs based on
the original URI returned by the Filer with ever shorter paths,
component by component. Semantically, I kept the old behavior of
scanning all the possible paths up until the filesystem root, so it
doesn't break projects that rely on this behavior.

### CODE REVIEW

File: processor/src/main/java/org/bindgen/processor/util/ConfUtil.java [hunk 1]
```diff
@@ -58,35 +56,58 @@
...
-		final String dummyPath;
-		try {
-			// We don't actually create this, we just want its URI
-			FileObject dummyFileObject = env.getFiler().getResource(location, "", "dummy.txt");
-			dummyPath = dummyFileObject.toUri().toString().replaceAll("file:", "");
-		} catch (IOException e1) {
-			return null;
-		}
+	private static InputStream resolveBindgenPropertiesIfExists(Location location, ProcessingEnvironment env, String fileName) {
+	    final String propertiesFilename = "bindgen.properties";
```
Comment (florianlacreuse) [inline processor/src/main/java/org/bindgen/processor/util/ConfUtil.java]: Why did you add and use `propertiesFilename` (constant)? Why not use parameter `fileName` as previously?

https://github.com/nfekete/bindgen/blob/07327351e85e6f064608cd7928cb4f17a0ebe7af/processor/src/main/java/org/bindgen/processor/util/ConfUtil.java#L75-L78

File: processor/src/main/java/org/bindgen/processor/util/ConfUtil.java [hunk 2]
```diff
@@ -58,35 +56,58 @@
...
+		final Optional<InputStream> bindingPropertiesInAncestorPaths = IntStream.iterate(dirPathComponents.size(), i -> i - 1)
+            .limit(dirPathComponents.size() - 1)
+            .boxed()
+            .map(i -> dirPathComponents.subList(0, i))
+            .map(list -> list.stream()
+                .collect(Collectors.joining("/", "/", "/" + propertiesFilename)))
+            .<Optional<InputStream>> map(newPath -> {
+                try {
+                    URL url = new URL(baseUrl, newPath);
+                    InputStream inputStream = url.openStream();
```
Comment (florianlacreuse) [inline processor/src/main/java/org/bindgen/processor/util/ConfUtil.java]: No `try-with-resources` for this `InputStream`? See https://github.com/igloo-project/bindgen/pull/13/files#diff-df832b7eb06ce65f926f349ea1e5c2e3R30 for example.

**Comment by florianlacreuse** [comment]
Hi @nfekete 

Thanks for your help and the PR. I'm not used to deal with bindgen source code but I will do my best to review your changes and merge it as quick as possible.

I already add a couple of comments, may be more to come (cf. conversations in _files changed_ tab).

Note: be careful to use tabs for indentation.

File: processor/src/main/java/org/bindgen/processor/util/ConfUtil.java (hunk 1 above)
Comment (nfekete) [reply processor/src/main/java/org/bindgen/processor/util/ConfUtil.java]: Good catch, thanks!

File: processor/src/main/java/org/bindgen/processor/util/ConfUtil.java (hunk 2 above)
Comment (nfekete) [reply processor/src/main/java/org/bindgen/processor/util/ConfUtil.java]: Try-with resources would close the stream at the (possibly implicit) _`finally`_ part of the try block, but we need to return an `InputStream` that is still open to use at the caller side.

**Comment by nfekete** [comment]
Thanks, I've seen that tab is the preferred indentation, and I thought that IDEA will respect the auto-detected indentation style, but I'll double check to be sure.

**Comment by nfekete** [comment]
I changed on the implementation a bit, as I wasn't really satisfied with the original fix. I've also added a test for the easily testable part of going over the possible parent paths of a path. I also made sure that indents are indeed tabs.

File: processor/src/main/java/org/bindgen/processor/util/ConfUtil.java [hunk 3]
```diff
@@ -19,18 +18,18 @@
...
 				break;
 			}
 		}
 
-		if (file != null) {
+		if (inputStream != null) {
 			Properties p = new Properties();
-			try {
-				p.load(new FileInputStream(file));
+			try (InputStream is = inputStream) {
```
Comment (florianlacreuse) [inline processor/src/main/java/org/bindgen/processor/util/ConfUtil.java:33]: Side note: with Java 9+ we could just use `inputStream` instead of using a new local variable (`is`). I will check with @lalmeras if we can upgrade it, any time soon.

File: processor/src/main/java/org/bindgen/processor/util/ConfUtil.java (hunk 2 above)
Comment (florianlacreuse) [reply processor/src/main/java/org/bindgen/processor/util/ConfUtil.java]: May be i'm a too paranoid but what if while computing `bindingPropertiesInAncestorPaths` we open multiple `InputStream`? Then only one will be used and closed properly.

See https://github.com/igloo-project/bindgen/pull/13/files#diff-df832b7eb06ce65f926f349ea1e5c2e3R78

File: processor/src/main/java/org/bindgen/processor/util/ConfUtil.java [hunk 4]
```diff
@@ -58,35 +57,60 @@
...
-		if (possible.exists()) {
-			return possible;
-		}
+	}
 
-		// No file found
-		return null;
+	static Stream<String> pathsToRoot(String path) {
+		final List<String> componentsWithFile = Arrays.asList(path.split("/"));
+		final List<String> pathComponents = componentsWithFile.subList(0, componentsWithFile.size());
```
Comment (florianlacreuse) [inline processor/src/main/java/org/bindgen/processor/util/ConfUtil.java]: The new implementation looks cleaner but what is the purpose of this line (sublist from 0 to size)? May be you first thought was to get ride of the `dummy.txt` here and this is a legacy?

File: processor/src/main/java/org/bindgen/processor/util/ConfUtil.java (hunk 3 above)
Comment (nfekete) [reply processor/src/main/java/org/bindgen/processor/util/ConfUtil.java:33]: About 80% of Java developers said in the largest [survey](https://snyk.io/blog/jvm-ecosystem-report-2018) about the JVM Ecosystem that their main projects are running on JDK 8. Keeping Java 8 compatibility is essential in my opinion.

File: processor/src/main/java/org/bindgen/processor/util/ConfUtil.java (hunk 4 above)
Comment (nfekete) [reply processor/src/main/java/org/bindgen/processor/util/ConfUtil.java]: 🤦‍♂️ Yes. Good catch, I'll remove it.

File: processor/src/main/java/org/bindgen/processor/util/ConfUtil.java (hunk 2 above)
Comment (nfekete) [reply processor/src/main/java/org/bindgen/processor/util/ConfUtil.java]: There are two cases:

- If the opening of the `InputStream` succeeds it should be returned right away wrapped in an `Optional`, so we open at most one `InputStream`s. The closing of the returned `InputStream` will be the responsibility of the caller, which is done in the [try-with-resources block](https://github.com/igloo-project/bindgen/pull/13/files#diff-df832b7eb06ce65f926f349ea1e5c2e3R31) in the caller method.
- In case the opening of the `InputStream` doesn't succeed, i.e. it throws an exception, we have nothing to close, as the exception itself got thrown while opening the stream, so there's no constructed `InputStream` instance to close. Not even try with resources would help in that case, as it's just a syntactic sugar around `AutoCloseable` instances.

Comment (florianlacreuse) [reply processor/src/main/java/org/bindgen/processor/util/ConfUtil.java]: Oh ok I got it. For a moment I forgot that thanks to `.findFirst()` we will stop immediately on the first good match. My bad, LGTM.

File: processor/src/main/java/org/bindgen/processor/util/ConfUtil.java [hunk 5]
```diff
@@ -58,35 +57,59 @@
...
 
-		// No file found
-		return null;
+	static Stream<String> pathsToRoot(String path) {
+		final List<String> pathComponents = Arrays.asList(path.split("/"));
+		final Stream<String> pathsToRoot = IntStream.iterate(pathComponents.size(), i -> i - 1)
+			.limit(pathComponents.size())
+			.boxed()
+			.map(i -> pathComponents.subList(0, i))
+			.filter(list -> !Arrays.asList("").equals(list)) //filter out empty path component before leading /
```
Comment (florianlacreuse) [inline processor/src/main/java/org/bindgen/processor/util/ConfUtil.java]:
```suggestion
			.filter(list -> !Arrays.asList("").equals(list)) // Filter out empty path component before leading /
```

File: processor/src/main/java/org/bindgen/processor/util/ConfUtil.java [hunk 6]
```diff
@@ -58,35 +57,59 @@
...
 		} catch (IOException e1) {
 			return null;
 		}
 
-		// Walk up looking for a bindgen.properties
-		File current = new File(dummyPath).getParentFile();
-		while (current != null) {
-			File possible = new File(current, fileName);
+		final Optional<InputStream> bindingPropertiesInAncestorPaths = pathsToRoot(baseUrl.getPath())
+			//skip the path that still contains the filename - we're gonna append that to every parent path anyway
```
Comment (florianlacreuse) [inline processor/src/main/java/org/bindgen/processor/util/ConfUtil.java:72]:
```suggestion
			// Skip the path that still contains the filename - we're gonna append that to every parent path anyway
```

**Comment by florianlacreuse** [comment]
@nfekete Just an idea but may be we could use `java.nio.Path` to deal with paths?

Something like this:

```java
Path basePath = Paths.get(fileObject.toUri());

// [...]

static Collection<Path> pathsToRoot(Path path) {
	if (path == null) {
		return new ArrayList<Path>();
	}
	List<Path> paths = new ArrayList<>(pathsToRoot(path.getParent()));
	paths.add(path);
	return paths;
}
```

(Use recursivity here but if you know how to list all parents with lambda i'm in.)

This snippet gives:
 1. `"/a/b/dummy.txt"` > `[/, /a, /a/b, /a/b/dummy.txt]`
 2. `"a/b/dummy.txt"` > `[a, a/b, a/b/dummy.txt]`

I don't know yet if the new `/` in case 1 is an issue. What do you think?

**Comment by nfekete** [comment]
I think that's a good idea. I didn't know about the pretty decent interplay between `Path` and `URI`. I'm guessing you did the above test on a posix system, where the path separator is `/` and one of the filesystem roots is `Path("/")`. On Windows, referencing a filesystem root is a bit different, which made the corresponding JUnit test a bit more involving as you can see. Overall I think this is a better approach so I suggest we use this version.

**Comment by florianlacreuse** [comment]
Thanks for your work! I will check with @lalmeras today for a final review with him. I hope we can merge this today.

**Review by florianlacreuse — APPROVED** [review]

**Comment by florianlacreuse** [comment]
@nfekete PR is approuved by @lalmeras. I will merge it monday. Thanks again for your help to maintain and fix bindgen.

**Comment by nfekete** [comment]
Thank you too for taking the time to review. There are other pending pull requests that solve other issues, btw. 😃

**Comment by florianlacreuse** [comment]
@nfekete We will check other PR as soon as possible.
